apps/nextjs/src/app/api/chat/chatHandler.ts

```diff
@@ -167,59 +171,227 @@ async function generateChatStream(
   );
 }
 
+function hasLessonPlan(obj: unknown): obj is { lessonPlan: unknown } {
+  return obj !== null && typeof obj === "object" && "lessonPlan" in obj;
+}
+
+function isValidLessonPlan(lessonPlan: unknown): boolean {
+  return lessonPlan !== null && typeof lessonPlan === "object";
+}
+
+function hasMessages(obj: unknown): obj is { messages: unknown } {
+  return obj !== null && typeof obj === "object" && "messages" in obj;
+}
+
+function isValidMessages(messages: unknown): boolean {
+  return Array.isArray(messages);
+}
+
+function verifyChatOwnership(
+  chat: { userId: string },
+  requestUserId: string,
+  chatId: string,
+): void {
+  if (chat.userId !== requestUserId) {
+    log.error(
+      `User ${requestUserId} attempted to access chat ${chatId} which belongs to ${chat.userId}`,
+    );
+    throw new Error("Unauthorized access to chat");
+  }
+}
+
+function parseChatOutput(
+  output: unknown,
+  chatId: string,
+): { messages: Message[]; lessonPlan: LooseLessonPlan } {
+  let messages: Message[] = [];
+  let lessonPlan: LooseLessonPlan = {};
+
+  try {
+    const parsedOutput =
+      typeof output === "string" ? JSON.parse(output) : output;
+
+    if (hasMessages(parsedOutput) && isValidMessages(parsedOutput.messages)) {
+      messages = parsedOutput.messages as Message[];
+    }
+
+    if (
+      hasLessonPlan(parsedOutput) &&
+      isValidLessonPlan(parsedOutput.lessonPlan)
+    ) {
+      lessonPlan = parsedOutput.lessonPlan as LooseLessonPlan;
+    }
+  } catch (error) {
+    log.error(`Error parsing output for chat ${chatId}`, error);
+  }
+
+  return { messages, lessonPlan };
+}
+
+async function loadChatDataFromDatabase(
+  chatId: string,
+  userId: string,
+): Promise<{ messages: Message[]; lessonPlan: LooseLessonPlan }> {
+  try {
+    const chat = await prisma.appSession.findUnique({
+      where: { id: chatId },
+      select: {
+        id: true,
+        userId: true,
+        output: true,
+      },
+    });
+
+    if (!chat) {
+      log.info(`No existing chat found for id: ${chatId}`);
+      return { messages: [], lessonPlan: {} };
+    }
+
+    verifyChatOwnership(chat, userId, chatId);
+
+    const { messages, lessonPlan } = parseChatOutput(chat.output, chatId);
+
+    log.info(
+      `Loaded ${messages.length} messages and lesson plan for chat ${chatId}`,
+    );
+    return { messages, lessonPlan };
+  } catch (error) {
+    log.error(`Error loading chat data for chat ${chatId}`, error);
+    return { messages: [], lessonPlan: {} };
+  }
+}
+
+function extractLatestUserMessage(frontendMessages: Message[]): Message | null {
+  if (!frontendMessages || frontendMessages.length === 0) {
+    return null;
+  }
+
+  for (let i = frontendMessages.length - 1; i >= 0; i--) {
+    const message = frontendMessages[i];
+    if (message && message.role === "user") {
+      return message;
+    }
+  }
+
+  return null;
```

Comment:
I think this can be tidied with [findLast](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findLast)

```suggestion
  return (frontendMessages || []).findLast(m => m.message?.role === "user")
```

Reply:
Thanks!